File: MISRA_C_vs_CPP/pilot/templates/java/Bench.java
import java.nio.file.*;
import java.nio.charset.StandardCharsets;

// Deliberately terse. Two ops the C engine does:
//   find <store> <substr>   full scan of the store, count records that match
//   and  <post1> <post2>    intersect two sorted posting lists (the AND of keys)
// Byte-level, no per-line String allocation, whole file pulled from page cache
// (warm) so we measure CPU+RAM like C's warm scan, not the disk.
public class Bench {
  static final int K = 7;                       // iterations: iter 0 = cold (JIT), rest warm

  public static void main(String[] a) throws Exception {
    if (a[0].equals("find")) {
      byte[] s = Files.readAllBytes(Paths.get(a[1]));
      byte[] p = a[2].getBytes(StandardCharsets.UTF_8);
      for (int it = 0; it < K; it++) {
        long t0 = System.nanoTime();
        int hits = countMatchingLines(s, p);
        System.out.printf("find  iter%d  %5.0f ms  hits=%d%n", it, (System.nanoTime()-t0)/1e6, hits);
      }
    } else { // and
      long[] A = readPostings(a[1]), B = readPostings(a[2]);
      for (int it = 0; it < K; it++) {
        long t0 = System.nanoTime();
        int c = intersect(A, B);
        System.out.printf("and   iter%d  %6.1f ms  common=%d  |A|=%d |B|=%d%n",
                          it, (System.nanoTime()-t0)/1e6, c, A.length, B.length);
      }
    }
  }

  static int countMatchingLines(byte[] s, byte[] p) {
    int hits = 0, pl = p.length, n = s.length; boolean line = false; byte p0 = p[0];
    for (int i = 0; i < n; i++) {
      byte b = s[i];
      if (b == '\n') { if (line) hits++; line = false; }
      else if (!line && b == p0 && i + pl <= n) {
        int j = 1; while (j < pl && s[i+j] == p[j]) j++;
        if (j == pl) line = true;
      }
    }
    if (line) hits++;
    return hits;
  }

  static long[] readPostings(String path) throws Exception {
    byte[] d = Files.readAllBytes(Paths.get(path));
    int lines = 0; for (byte b : d) if (b == '\n') lines++;
    long[] out = new long[lines]; int k = 0; long cur = 0; boolean any = false;
    for (byte b : d) {
      if (b == '\n') { if (any) out[k++] = cur; cur = 0; any = false; }
      else if (b >= '0' && b <= '9') { cur = cur*10 + (b - '0'); any = true; }
    }
    if (k != out.length) { long[] t = new long[k]; System.arraycopy(out,0,t,0,k); out = t; }
    return out; // idx posting lists are already sorted ascending
  }

  static int intersect(long[] A, long[] B) {
    int i=0,j=0,c=0;
    while (i<A.length && j<B.length) {
      long x=A[i], y=B[j];
      if (x==y){c++;i++;j++;} else if (x<y) i++; else j++;
    }
    return c;
  }
}

File: MISRA_C_vs_CPP/pilot/templates/rust/bench.rs
// Same two ops as the C/Ada/Java/Python benches. Safe idiomatic Rust: indexing
// is bounds-checked at run time (the optimizer elides most in these loops), the
// slice compare is a memcmp. Build: rustc -O bench.rs -o bench_rust
use std::env;
use std::fs;
use std::time::Instant;

fn count_lines(s: &[u8], pat: &[u8]) -> usize {
    let (mut hits, mut line) = (0usize, false);
    let (pl, p0, n) = (pat.len(), pat[0], s.len());
    let mut i = 0;
    while i < n {
        let b = s[i];
        if b == b'\n' {
            if line { hits += 1; }
            line = false;
        } else if !line && b == p0 && i + pl <= n && &s[i..i + pl] == pat {
            line = true;
        }
        i += 1;
    }
    if line { hits += 1; }
    hits
}

fn read_postings(path: &str) -> Vec<i64> {
    let d = fs::read(path).unwrap();
    let mut out = Vec::new();
    let (mut cur, mut any) = (0i64, false);
    for &b in &d {
        if b == b'\n' {
            if any { out.push(cur); }
            cur = 0; any = false;
        } else if b.is_ascii_digit() {
            cur = cur * 10 + (b - b'0') as i64;
            any = true;
        }
    }
    out // idx posting lists are already sorted ascending
}

fn intersect(a: &[i64], b: &[i64]) -> usize {
    let (mut i, mut j, mut c) = (0usize, 0usize, 0usize);
    while i < a.len() && j < b.len() {
        let (x, y) = (a[i], b[j]);
        if x == y { c += 1; i += 1; j += 1; }
        else if x < y { i += 1; }
        else { j += 1; }
    }
    c
}

fn main() {
    let a: Vec<String> = env::args().collect();
    let k = 7;
    if a[1] == "find" {
        let s = fs::read(&a[2]).unwrap();
        let pat = a[3].as_bytes();
        for it in 0..k {
            let t = Instant::now();
            let h = count_lines(&s, pat);
            println!("find  iter{}  {:5.0} ms  hits={}", it, t.elapsed().as_secs_f64() * 1e3, h);
        }
    } else {
        let (pa, pb) = (read_postings(&a[2]), read_postings(&a[3]));
        for it in 0..k {
            let t = Instant::now();
            let c = intersect(&pa, &pb);
            println!("and   iter{}  {:6.1} ms  common={} |A|={} |B|={}",
                     it, t.elapsed().as_secs_f64() * 1e3, c, pa.len(), pb.len());
        }
    }
}
